In [1]:
import spacy
nlp = spacy.load('en_core_web_sm')
from spacy.matcher import Matcher
import pandas as pd
df = pd.read_csv(r"C:/Users/uchei/OneDrive/Desktop/Gozie/Raw_Skills_Dataset.csv")

In [2]:
df.head()

RAW DATA
0         What ifs
1        seniority
2      familiarity
3  functionalities
4          Lambdas

In [3]:
df.shape

(34116, 1)

There are 34116 samples in the raw data. Let's filter out the useful information.

In [4]:
def patterns(text):
    doc = nlp(text)
    patterns = []
    for token in doc:
        patterns.append(token.pos_)
    return patterns

Let's extract the patterns from the example dataset.

In [5]:
examples = pd.read_csv(r"C:/Users/uchei/OneDrive/Desktop/Gozie/Example_Technical_Skills.csv")
examples.head()

Technology Skills
0                    SAP Fiori Developer
1  Oracle Instance Management & Strategy
2           Boomi Master Data Management
3  Digital Manufacturing on Cloud ( DMC)
4                                 DevOps

In [6]:
examples['Patterns'] = examples['Technology Skills'].apply(lambda x:patterns(x))

In [7]:
examples.head()

Technology Skills  \
0                    SAP Fiori Developer   
1  Oracle Instance Management & Strategy   
2           Boomi Master Data Management   
3  Digital Manufacturing on Cloud ( DMC)   
4                                 DevOps   

                                          Patterns  
0                            [PROPN, PROPN, PROPN]  
1              [PROPN, PROPN, PROPN, CCONJ, PROPN]  
2                     [PROPN, PROPN, PROPN, PROPN]  
3  [PROPN, PROPN, ADP, PROPN, PUNCT, PROPN, PUNCT]  
4                                          [PROPN]

Using these defined patterns and finding similarities, we can create our own Rule-Based POS Matching

In [8]:
def extract_patterns(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) ==1:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [9]:
one_word= examples['Technology Skills'].apply(lambda x:extract_patterns(x))

In [10]:
one_word = one_word[one_word.apply(len) > 0]
one_word= one_word.apply(lambda x: ','.join(map(str,x)))

In [11]:
one_word.unique()

array(['PROPN', 'INTJ', 'NOUN', 'VERB', 'ADJ', 'PROPN,PUNCT,NOUN', 'ADV',
       'PROPN,SYM,NOUN', 'VERB,PUNCT,VERB'], dtype=object)

In [12]:
def extract_two_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 2:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [13]:
two_words = examples['Technology Skills'].apply(lambda x:extract_two_words(x))
two_words = two_words[two_words.apply(len) > 0]
two_words= two_words.apply(lambda x: ','.join(map(str,x)))

In [14]:
two_words.unique()

array(['PROPN,PROPN', 'ADJ,ADJ', 'NOUN,VERB', 'PROPN,NOUN', 'NOUN,NUM',
       'PROPN,PROPN,SYM,PROPN,SYM,PROPN', 'VERB,PROPN,SYM,PROPN',
       'NOUN,NOUN', 'PROPN,PUNCT,PROPN,PUNCT', 'ADJ,NOUN', 'PROPN,VERB',
       'INTJ,NOUN', 'NUM,NOUN', 'NOUN,PROPN', 'ADJ,PROPN',
       'PROPN,ADJ,ADJ,NOUN', 'ADP,NOUN', 'ADJ,VERB',
       'PROPN,PROPN,PUNCT,PROPN', 'PROPN,ADP', 'VERB,NOUN', 'VERB,VERB',
       'PROPN,PUNCT', 'VERB,PROPN', 'NOUN,PUNCT,ADP,PUNCT,NOUN,NOUN',
       'PROPN,SYM,PROPN,PROPN', 'NOUN,SYM,PROPN,PROPN',
       'PROPN,SYM,NOUN,PUNCT,NOUN,NOUN', 'PUNCT,PROPN',
       'PROPN,PUNCT,PROPN,PROPN'], dtype=object)

In [15]:
def extract_three_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 3:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [16]:
three_words = examples['Technology Skills'].apply(lambda x:extract_three_words(x))
three_words = three_words[three_words.apply(len) > 0]
three_words= three_words.apply(lambda x: ','.join(map(str,x)))

In [17]:
three_words.unique()

array(['PROPN,PROPN,PROPN', 'PROPN,SYM,PROPN', 'PROPN,PROPN,NOUN',
       'PROPN,PROPN,PUNCT,PROPN,PROPN', 'NOUN,CCONJ,NOUN',
       'NOUN,NOUN,PUNCT,PROPN,PUNCT', 'PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'ADJ,PROPN,PUNCT,PROPN,PUNCT', 'NOUN,ADP,NOUN', 'NOUN,PROPN,NOUN',
       'NOUN,PROPN,PROPN', 'PROPN,PROPN,VERB', 'ADJ,NOUN,NOUN',
       'PROPN,NOUN,NOUN', 'PROPN,PROPN,NOUN,PUNCT', 'PROPN,PUNCT,ADJ',
       'PROPN,PUNCT,PROPN', 'PROPN,PROPN,PROPN,SYM,PROPN',
       'PROPN,PROPN,SYM,PROPN,PROPN', 'PROPN,ADJ,NOUN,PUNCT,ADP',
       'PROPN,PROPN,NUM,NOUN', 'NOUN,PROPN,PROPN,PUNCT,PROPN',
       'NOUN,NOUN,PROPN', 'NOUN,PUNCT,PROPN,PROPN,PROPN,PUNCT',
       'PROPN,ADP,NOUN', 'PROPN,PUNCT,PROPN,PROPN', 'PROPN,PROPN,NUM',
       'PROPN,SYM,PROPN,PROPN,PROPN', 'NOUN,VERB,NOUN',
       'PROPN,PROPN,PUNCT,NOUN,PUNCT', 'PROPN,NOUN,PROPN',
       'NOUN,PUNCT,PROPN,PROPN', 'NOUN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,ADJ,NOUN', 'PROPN,PROPN,PROPN,PUNCT,PROPN',
       'NOUN,PUNCT,PROPN', 'PROPN

In [18]:
def extract_four_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 4:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [19]:
four_words = examples['Technology Skills'].apply(lambda x:extract_four_words(x))
four_words = four_words[four_words.apply(len) > 0]
four_words= four_words.apply(lambda x: ','.join(map(str,x)))

In [20]:
four_words.unique()

array(['PROPN,PROPN,PROPN,PROPN', 'NOUN,PUNCT,PROPN,PROPN',
       'PROPN,PROPN,SYM,PROPN', 'PROPN,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'NOUN,PROPN,PROPN,NOUN', 'NOUN,PUNCT,NOUN,PROPN,PUNCT,PROPN',
       'NOUN,PROPN,PROPN,PROPN', 'PROPN,PROPN,CCONJ,PROPN',
       'NOUN,ADP,NOUN,NOUN', 'PROPN,PROPN,NOUN,VERB',
       'PROPN,PROPN,PUNCT,PROPN', 'PROPN,PROPN,PUNCT,PROPN,PROPN,PUNCT',
       'NUM,PROPN,PROPN,SYM,PROPN,PROPN,PROPN', 'PROPN,PROPN,NOUN,NOUN',
       'PROPN,PROPN,PUNCT,NOUN,NOUN,NOUN', 'PROPN,PUNCT,PROPN,NOUN',
       'PROPN,VERB,NOUN,NOUN', 'PROPN,PUNCT,PROPN,PROPN',
       'PROPN,PROPN,VERB,NOUN', 'PROPN,PROPN,PROPN,PUNCT,NOUN,PUNCT',
       'PROPN,SYM,PROPN,PROPN,SYM,PROPN,PROPN,PROPN',
       'PROPN,PROPN,PROPN,VERB,PUNCT,ADP', 'PROPN,NUM,NUM,NOUN',
       'PROPN,PROPN,PROPN,VERB', 'PROPN,PROPN,VERB,ADJ',
       'PROPN,PUNCT,NOUN,PROPN,PUNCT,PROPN', 'PROPN,PROPN,SYM,NOUN',
       'PROPN,PUNCT,ADJ,ADJ,ADJ,NOUN', 'PROPN,PROPN,PROPN,NOUN',
       'PROPN,PUNCT,PROPN,PROPN,S

In [21]:
def extract_five_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 5:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [22]:
five_words = examples['Technology Skills'].apply(lambda x:extract_five_words(x))
five_words = five_words[five_words.apply(len) > 0]
five_words= five_words.apply(lambda x: ','.join(map(str,x)))
five_words.unique()

array(['PROPN,PROPN,PROPN,CCONJ,PROPN',
       'PROPN,PROPN,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,PROPN,PROPN,PROPN',
       'PROPN,AUX,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'NOUN,PROPN,PROPN,PROPN,PROPN',
       'NOUN,SYM,PROPN,PUNCT,PROPN,CCONJ,NOUN',
       'PROPN,PUNCT,PROPN,PROPN,PROPN', 'PROPN,ADJ,PROPN,PROPN,NOUN',
       'PROPN,ADJ,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,CCONJ,PROPN,PROPN', 'PROPN,PROPN,PUNCT,PROPN,PROPN',
       'NOUN,PUNCT,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,PUNCT,PROPN,PROPN,PROPN,PUNCT',
       'PROPN,SYM,PROPN,PROPN,CCONJ,PROPN,PROPN',
       'PROPN,NOUN,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'NOUN,VERB,PROPN,PROPN,PROPN', 'NOUN,PROPN,PUNCT,PROPN,PROPN',
       'NOUN,PROPN,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,PROPN,PROPN,PUNCT,NOUN,PUNCT',
       'PROPN,PROPN,NOUN,NOUN,NOUN', 'NOUN,NOUN,ADP,PROPN,NUM',
       'PROPN,PROPN,PROPN,ADP,NOUN',
       'PROPN,PROPN,PUNCT,PROPN,PUNCT,PROPN,PROPN',
       'P

In [23]:
def extract_six_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 6:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [24]:
six_words = examples['Technology Skills'].apply(lambda x:extract_six_words(x))
six_words = six_words[six_words.apply(len) > 0]
six_words= six_words.apply(lambda x: ','.join(map(str,x)))
six_words.unique()

array(['PROPN,PROPN,ADP,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PUNCT,PROPN,NOUN,PROPN,PROPN',
       'PROPN,PROPN,PROPN,PROPN,PROPN,PROPN',
       'PROPN,PROPN,NOUN,PROPN,PROPN,PROPN',
       'PROPN,PROPN,PROPN,CCONJ,PROPN,NOUN',
       'NOUN,NOUN,PUNCT,NOUN,CCONJ,NOUN',
       'PROPN,PUNCT,PROPN,NOUN,NOUN,NOUN',
       'PROPN,PROPN,PROPN,PUNCT,NOUN,PUNCT,PROPN,X,PUNCT',
       'PROPN,PROPN,NOUN,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,PROPN,PROPN,PUNCT,PROPN',
       'PROPN,PUNCT,NOUN,PUNCT,NOUN,SYM,PROPN,PUNCT,PROPN,CCONJ,PROPN',
       'PROPN,PROPN,PROPN,PROPN,PROPN,PUNCT,PROPN,PROPN,PROPN,PUNCT',
       'ADJ,SYM,ADJ,SYM,ADJ,NOUN',
       'PROPN,PUNCT,PROPN,PUNCT,PROPN,PUNCT,PROPN,CCONJ,PROPN',
       'PROPN,PROPN,PROPN,PROPN,CCONJ,PROPN',
       'PROPN,PROPN,CCONJ,PROPN,PUNCT,PROPN,PUNCT,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,NUM,PUNCT,PROPN,NOUN',
       'PROPN,PROPN,PROPN,PROPN,PROPN,PUNCT,PROPN,PUNCT',
       'PROPN,PROPN,PROPN,CCONJ,PROPN,PROPN',
       'PROPN,PROP

In [25]:
def extract_seven_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 7:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [26]:
seven_words = examples['Technology Skills'].apply(lambda x:extract_seven_words(x))
seven_words = seven_words[seven_words.apply(len) > 0]
seven_words= seven_words.apply(lambda x: ','.join(map(str,x)))
seven_words.unique()

array(['PROPN,PROPN,PROPN,PROPN,ADP,PROPN,PUNCT,PROPN',
       'PROPN,VERB,PUNCT,PROPN,PROPN,CCONJ,VERB',
       'NOUN,PROPN,PROPN,PROPN,PROPN,PROPN,PROPN',
       'PROPN,PROPN,PROPN,NUM,PUNCT,ADJ,PROPN',
       'PROPN,PROPN,PUNCT,PROPN,PROPN,PUNCT,CCONJ,NOUN'], dtype=object)

In [27]:
def extract_eight_words(text):
    doc = nlp(text)
    patterns = []
    if len(text.split()) == 8:
        for token in doc:
            patterns.append(token.pos_)
    return patterns

In [28]:
eight_words = examples['Technology Skills'].apply(lambda x:extract_eight_words(x))
eight_words = eight_words[eight_words.apply(len) > 0]
eight_words= eight_words.apply(lambda x: ','.join(map(str,x)))
eight_words.unique()

array([], dtype=object)

The maximum number of words contained in each row in the example is seven. We will use these unique patterns to extract the important information.

In [33]:
def extract_tech(text):
    doc = nlp(text)
    skills = []
    if len(text.split()) == 1:
        pattern = [[{'POS': 'PROPN'}],
                   [{'POS': 'INTJ'}],
                   [{'POS': 'VERB'}],
                   [{'POS': 'ADJ'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}],
                   [{'POS': 'ADV'}],
                   [{'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'NOUN'}],
                   [{'POS': 'VERB'}, {'POS': 'PUNCT'}, {'POS': 'VERB'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_1', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
            
    elif len(text.split()) == 2:
        pattern = [[{'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'ADJ'}, {'POS': 'ADJ'}],
                   [{'POS': 'NOUN'}, {'POS': 'VERB'}],
                   [{'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'NUM'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}],
                   [{'POS': 'VERB'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'VERB'}],
                   [{'POS': 'INTJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'NUM'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}],
                   [{'POS': 'ADJ'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'ADJ'}, {'POS': 'ADJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'ADP'}, {'POS': 'NOUN'}],
                   [{'POS': 'ADJ'}, {'POS': 'VERB'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'ADP'}],
                   [{'POS': 'VERB'}, {'POS': 'NOUN'}],
                   [{'POS': 'VERB'}, {'POS': 'VERB'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'VERB'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'ADP'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_2', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
            
    elif len(text.split()) == 3:
        pattern = [[{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NUM'}],
                   [{'POS': 'NOUN'}, {'POS': 'CCONJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'ADJ'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'ADP'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'VERB'}],
                   [{'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'VERB'}, {'POS': 'NOUN'}],
                   [{'POS': 'VERB'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'ADV'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'VERB'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'X'}],
                   [{'POS': 'PROPN'}, {'POS': 'ADJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'},{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'},{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_3', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
            
    elif len(text.split()) == 4:
        pattern = [[{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'VERB'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'ADP', 'OP': '?'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'VERB'}, {'POS': 'NOUN'}],
                   [{'POS': 'ADJ'}, {'POS': 'AUX'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}], 
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'NUM'}, {'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_4', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
            
    elif len(text.split()) == 5:
        pattern = [[{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'AUX'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'VERB'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}, {'POS': 'NUM'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'ADJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'ADJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}]]
                                                                                                              
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_5', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
            
    elif len(text.split()) == 6:
        pattern = [[{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NUM'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'VERB'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'NUM'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'ADJ'}, {'POS': 'SYM'}, {'POS': 'ADJ'}, {'POS': 'SYM'}, {'POS': 'ADJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'ADJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'CCONJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'SYM'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'X'}, {'POS': 'PUNCT'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PUNCT'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NUM'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'NOUN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'NOUN'}],
                   [{'POS': 'ADJ'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'CCONJ'}, {'POS': 'NOUN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS':'PROPN'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_6', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
    
    elif len(text.split()) == 7:
        pattern = [[{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'ADP'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'VERB'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'CCONJ'}, {'POS': 'VERB'}],
                   [{'POS': 'NOUN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'NUM'}, {'POS': 'PUNCT'}, {'POS': 'ADJ'}, {'POS': 'PROPN'}],
                   [{'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'PROPN'}, {'POS': 'PROPN'}, {'POS': 'PUNCT'}, {'POS': 'CCONJ'}, {'POS': 'NOUN'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add('rule_7', pattern, on_match = None)
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            skills.append(span.text)
    return skills

In [30]:
def soft_skills(text):
    doc = nlp(text)
    soft_skills = []
    if len(text.split()) == 1:
        pattern = [{'POS': 'NOUN'}]
        matcher = Matcher(nlp.vocab)
        matcher.add("soft", [pattern])
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            soft_skills.append(span.text)
    elif len(text.split()) == 2:
        pattern = [[{'POS': 'ADJ'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'NOUN'}]]
        matcher = Matcher(nlp.vocab)
        matcher.add("soft", [pattern[0]])
        matcher.add("soft", [pattern[1]])
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            soft_skills.append(span.text)
    elif len(text.split()) == 3:
        pattern = [[{'POS':'NOUN'},{'POS': 'PART'}, {'POS': 'VERB'}],
                    [{'POS': 'NOUN'}, {'POS': 'NOUN'}, {'POS': 'NOUN'}],
                   [{'POS': 'NOUN'}, {'POS': 'ADJ'}, {'POS': 'NOUN'}]]
                    
        matcher = Matcher(nlp.vocab)
        matcher.add("soft", [pattern[0]])
        matcher.add("soft", [pattern[1]])
        matcher.add("soft", [pattern[2]])
        matches = matcher(doc)
        for match_id, start, end in matches:
            span = doc[start:end]
            soft_skills.append(span.text)
    return soft_skills

In [31]:
Soft_Skills = df['RAW DATA'].apply(lambda x:soft_skills(x))

In [34]:
Technical_Skills = df['RAW DATA'].apply(lambda x: extract_tech(x))

In [35]:
#eliminate empty values
Technical_Skills = Technical_Skills[Technical_Skills.apply(len) > 0]
Technical_Skills = Technical_Skills.apply(lambda x: ','.join(map(str,x)))
len(Technical_Skills)

12546

There are 12546 technical skills according to the defined patterns derived from the example dataset.

In [36]:
Technical_Skills.head()

5             Java Streams
7     Relational Databases
8                      SQL
11               Hibernate
12                 MyBatis
Name: RAW DATA, dtype: object

In [37]:
#eliminate empty values
Soft_Skills = Soft_Skills[Soft_Skills.apply(len) > 0]
Soft_Skills = Soft_Skills.apply(lambda x: ','.join(map(str,x)))
len(Soft_Skills)

12899

Using the patterns defined, there are 12899 soft skills contained in the dataset. Let's preview some.

In [38]:
Soft_Skills.head()

1          seniority
2        familiarity
3    functionalities
4            Lambdas
9                ORM
Name: RAW DATA, dtype: object

There are some anomalies in the soft skills extracted, as well as some omitted from the technical skills dataset. This is owing to the duality of the singular POS tag NOUN, which can appear as both a technical skill and soft skill. However, the singular NOUN tag predominantly contains soft skills, hence the grouping into the soft skills category.